### PR TITLE
Pr/19

### PR DESCRIPTION
**Pull request by QhenryQ**

### CODE REVIEW

**Comment by QhenryQ** [comment]
@olenaboiko303 could you please help me merging this branch? This is the what we need for the manuscript, such visualization and parameter recovery.

**Comment by olenaboiko303** [comment]
> @olenaboiko303 could you please help me merging this branch? This is the what we need for the manuscript, such visualization and parameter recovery.

1.  Why do we need the notebooks `notebooks/annual_seasonality_train.ipynb` and `notebooks/old_model_train.ipynb` ? They are redundant and outdated, given that we have the most recent Python script to execute the models (both spatial and seasonality) in `nteprsm/model.py` and `nteprsm/model_umn.py`? If we want to keep these notebooks - they need to be updated, for example `old_model_train.ipynb` assumes the old versions of the variable names (I, J, P, M, etc)
2. Is this function https://github.com/QhenryQ/nteprsm/blob/494e0856fd6fdc39f5ff7c7031e3550e101e2e4c/notebooks/utils.py#L48C1-L48C16 different from https://docs.scipy.org/doc/scipy-1.15.3/reference/generated/scipy.special.softmax.html? If they are the same, then importing a function from a package might be better than creating a new custom function. And it actually seems that you used it already in a different file: https://github.com/QhenryQ/nteprsm/blob/494e0856fd6fdc39f5ff7c7031e3550e101e2e4c/nteprsm/utils.py#L16
3. ` rsm_probability` function appears multiple times: in https://github.com/QhenryQ/nteprsm/blob/pr/19/nteprsm/utils.py, https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/utils.py, and https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/nj2_viz.ipynb.
4. can `rsm_probability` and `rsm_probability_vector` functions be combined into one? 
5. does this notebook run: https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/nj2_viz.ipynb? Should it be committed with outputs? currently there are errors and outputs are empty. 
6. while I understand that `utils.py` in `notebooks` and `utils.py` in `nteprsm` serve different purposes, it's somewhat confusing to have the same filename for different files.

**Comment by QhenryQ** [comment]
1. I don't think we need it now.
2. probably not. Please double check on me tho. 
3. I am gonna rely on your judgement to get rid of the redundancy. 
4. if possible. we should vectorize as much as we can. could you please investigate?
5. it probably doesn't run, neither do some of other notebooks. I was using resources from my wife's company, so some of them were copied directly from the company's internal system especially for those need a lot computation. I need your help to make sure they all ran properly. 
6. consider change the one in notebooks to `rutils`?

The ultimate objective is to clean up this repo, remove redundancy as much as we can. I want to share these with other turf students who are capable of coding.

**Comment by olenaboiko303** [comment]
to re-run the following Notebooks https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/parameter_recovery_training.ipynb and https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/parameter_recovery_analysis.ipynb do we need to re-run a full model with "fake" data generated for synthetic 21x21 grid? Would that be better done outside of the Notebooks?

**Comment by QhenryQ** [comment]
> > yeah, I have seen issues like this before. It has something to do with the implementation of the model. Histrically, we had two different methods to keep track of entries across different rating events. The one using cumulative counts of entry could cause error like this. It seems unstable. You will need the following function to extract time_effect.
> > ```
> > def extract_time_effect(datahandler: "utils.DataHandler", fit: "stan fit object") -> pd.DataFrame:
> >     """
> >     Extract and format the time effect from a fitted model object.
> > 
> >     Args:
> >         datahandler (utils.DataHandler): Data handler containing model data with 'entry_cumcount' and 'adj_day_of_year' columns.
> >         fit: Fitted model object containing the 'time_effect' variable.
> > 
> >     Returns:
> >         pd.DataFrame: DataFrame with 'adj_day_of_year' as index and time effects as columns, sorted and formatted.
> >     """
> >     raing_event2doy = datahandler.model_data[['adj_day_of_year', 'rating_event_code']].drop_duplicates(
> >         ).sort_values(by='rating_event_code')
> >     time_effect = pd.DataFrame(
> >         fit.stan_variable("time_effect").mean(axis=0),
> >         columns=raing_event2doy['adj_day_of_year'].values
> >     )
> >     time_effect.index = time_effect.index + 1
> >     time_effect = time_effect.T.sort_index()
> >     time_effect.index.name = 'adj_day_of_year'
> >     return time_effect
> > ```
> 
> Now getting shape mismatch when trying to plug in this function: image
> 
> image

I believe you didn't pull the latest code. 
Screenshot 2025-06-25 at 2 35 40 AM

**Comment by QhenryQ** [comment]
> > > > yeah, I have seen issues like this before. It has something to do with the implementation of the model. Histrically, we had two different methods to keep track of entries across different rating events. The one using cumulative counts of entry could cause error like this. It seems unstable. You will need the following function to extract time_effect.
> > > > ```
> > > > def extract_time_effect(datahandler: "utils.DataHandler", fit: "stan fit object") -> pd.DataFrame:
> > > >     """
> > > >     Extract and format the time effect from a fitted model object.
> > > > 
> > > >     Args:
> > > >         datahandler (utils.DataHandler): Data handler containing model data with 'entry_cumcount' and 'adj_day_of_year' columns.
> > > >         fit: Fitted model object containing the 'time_effect' variable.
> > > > 
> > > >     Returns:
> > > >         pd.DataFrame: DataFrame with 'adj_day_of_year' as index and time effects as columns, sorted and formatted.
> > > >     """
> > > >     raing_event2doy = datahandler.model_data[['adj_day_of_year', 'rating_event_code']].drop_duplicates(
> > > >         ).sort_values(by='rating_event_code')
> > > >     time_effect = pd.DataFrame(
> > > >         fit.stan_variable("time_effect").mean(axis=0),
> > > >         columns=raing_event2doy['adj_day_of_year'].values
> > > >     )
> > > >     time_effect.index = time_effect.index + 1
> > > >     time_effect = time_effect.T.sort_index()
> > > >     time_effect.index.name = 'adj_day_of_year'
> > > >     return time_effect
> > > > ```
> > > 
> > > 
> > > Now getting shape mismatch when trying to plug in this function: image
> > > image
> > 
> > 
> > I believe you didn't pull the latest code. Screenshot 2025-06-25 at 2 35 40 AM
> 
> I have, the update is specifically for stan model code, not seeing how it is used in the visualization Notebooks. Do I need to re-run some or all models with this update?

You can re-run the model, using new data and new model file. or you can also use the fit objects saved the google drive.

**Comment by QhenryQ** [comment]
Could you please use the new data from google drive and the new model (https://github.com/QhenryQ/nteprsm/blob/pr/19/models/annual_seasonality_model.stan) to re-run model for the notebook? This will be a good opportunity for you to understand how the model runs under the hood. I am in a conference right now. Will try to get back to you as much as I can.

**Comment by QhenryQ** [comment]
If necessary, re-write the notebooks or whatever code that you believe is necessary to create the figures needed in the manuscript.

**Comment by olenaboiko303** [comment]
> Could you please use the new data from google drive and the new model (https://github.com/QhenryQ/nteprsm/blob/pr/19/models/annual_seasonality_model.stan) to re-run model for the notebook? This will be a good opportunity for you to understand how the model runs under the hood. I am in a conference right now. Will try to get back to you as much as I can.

Yes, I can certainly re-run the model and then the Notebooks. That's what I've been trying to do before I can start optimizing and reducing redundancy across the code base. But I keep running into errors.

Could you explain what you meant by `num_rating_events` you introduced in https://github.com/QhenryQ/nteprsm/pull/22/commits/4686338a5dacfbb4900fca50c831aec322edae17? This variable was NOT declared in the `dataparser` in the utilities. So my initial attempt resulted in the following error:
image

I tried to add the following to `stan_data` dictionary in utilities : `"num_rating_events": int(self.model_data.rating_event_code.max())`, but then I get a mismatch in dimensions error. Are `"num_ratings"` and `"num_rating_events"` different? Is the problem still in STAN code?

image

Did the model run on your side after you introduced your most recent edits?

**Comment by QhenryQ** [comment]
which `DataHandler` class did you use? variable `num_ratings` is defined [here](https://github.com/QhenryQ/nteprsm/blob/4686338a5dacfbb4900fca50c831aec322edae17/nteprsm/utils.py#L316). I don't think you need to add it manually. `time` variable is also defined and declared [here](https://github.com/QhenryQ/nteprsm/blob/4686338a5dacfbb4900fca50c831aec322edae17/nteprsm/utils.py#L332).

**Review by QhenryQ — COMMENTED** [review]
try this.

**Comment by QhenryQ** [comment]
`num_ratings` and `num_rating_events` are the same thing. We just have too many naming changes over the course of model development.

**Comment by olenaboiko303** [comment]
@QhenryQ 

- I have re-run the model with the most recent version of the seasonality model code (https://github.com/QhenryQ/nteprsm/blob/pr/19/models/annual_seasonality_model.stan)
- I have added a Notebook to derive seasonality curves: https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/nj2_viz_SEASONALITY.ipynb. While I’m not seeing the same noisy pattern I showed earlier, the curves are still not the same as what you included in the manuscript draft. So, I’m trying to understand: are the graphics in the manuscript just placeholders and we need to replace them? Or are they correct and I need to derive the exact same plots from the Notebooks? 
- Note that I made changes to the original code, including the function `extract_time_effect` you shared earlier in this PR discussion. So please review the code and let me know what issues you see. I intentionally kept only the code needed for a minimal reproducible example for seasonality curves specifically, so it’s easier to review. 
- I used the `svg` renderer to load the `plotly` plot in a static format, so it renders on GitHub.

**Comment by QhenryQ** [comment]
Good job on the get it to run! There are a lot of stuff in it. I will ran the code myself tomorrow to see what happened. The plots in the manuscript should be the correct one, as I was able to re-create the same plot.

**Comment by QhenryQ** [comment]
Just confirm your plot will be the final version. The original plot was generated from a different version of the parsed data, which has included more data in the 2nd parsing. We will stick to what we currently have. That's is why we need to consolidate the visualization notebooks.

**Comment by olenaboiko303** [comment]
- Based on our earlier discussion, I changed `utils.py` under notebooks to `rutils.py` to distinguish better from `nterprsm/utils.py`, Currently in your notebooks you still have `import utils as notebook_utils`, which won't work right now.  I used your pre-trained `pkl` objects for parameter recovery to run https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/parameter_recovery_analysis.ipynb and everything runs as expected after adjusting the imports.
- Does the last seasonality curve in https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/parameter_recovery_analysis.ipynb  (entry 94) look alright? 
- What about the remaining notebooks https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/pymc_visualization.ipynb and [visulization_demo_for_frontend_dev.ipynb](https://github.com/QhenryQ/nteprsm/blob/pr/19/notebooks/visulization_demo_for_frontend_dev.ipynb)? Do those need to be reviewed as well?

**Comment by QhenryQ** [comment]
1. could you please help me change them?
2. nice catch! entry 94 is correct, tho looks strange. It is to demonstrate what the fitted line looks like when the MAE is 0.73.
3. delete `notebooks/pymc_visualization.ipynb`
4. keep `notebooks/visulization_demo_for_frontend_dev.ipynb`

**Review by olenaboiko303 — APPROVED** [review]
